[FLINK-19276][json][connector-kafka] Support reading Debezium metadata

twalthr opened the pull request:
## What is the purpose of the change

This exposes metadata for the Debezium JSON format according to FLIP-107.

Note: A complete end-to-end test for Kafka + Debezium is missing yet.

## Brief change log

- Update the Kafka connector to expose format specific metadata.
- Reconfigure the internal JsonRowDataDeserializationSchema to read additional fields.
- Let DebeziumJsonDeserializationSchema access and convert those additional fields to metadata columns.

## Verifying this change

This change added tests and can be verified as follows: `DebeziumJsonSerDeSchemaTest`, `KafkaDynamicTableFactoryTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

wuchong commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java:
```diff
@@ -179,13 +183,37 @@ public ScanRuntimeProvider getScanRuntimeProvider(ScanContext context) {
 	public Map<String, DataType> listReadableMetadata() {
 		final Map<String, DataType> metadataMap = new LinkedHashMap<>();
+
+		// add value format metadata with prefix
```
Add a comment to mention why we must put the format metadata before connector metadata? I think the order can't be switched.

twalthr replied:
actually the order doesn't matter here, but I just noticed that the order should be the other way around if format metadata should have higher precedence

wuchong commented on flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaChangelogTableITCase.java:
```diff
@@ -131,7 +133,8 @@ public void testKafkaDebeziumChangelogSource() throws Exception {
 		TableResult tableResult = tEnv.executeSql(
-			"INSERT INTO sink SELECT name, SUM(weight) FROM debezium_source GROUP BY name");
+			"INSERT INTO sink SELECT FIRST_VALUE(origin), name, SUM(weight) "
+				+ "FROM debezium_source GROUP BY name");
```
Could you make this case more complex? For example, declaring 2 format metadata and 2 connector metadata, but only select 1 format metadata and 1 connector metadata. This helps to verify format and connector metadata can work together and the metadata pruning also works.

twalthr replied:
Good suggestion. I found a bug.

wuchong reviewed (commented):
Thanks for the great work @twalthr . I think the pull request has been in a good shape. 

I just find that the planner will prune metadata columns only when the source supports projection pushdown. Therefore, currently, the `applyReadableMetadata` method in Kafka source will never be called. Should we improve this?

twalthr commented:
Thanks for the feedback @wuchong. I updated the PR and hope that I could address your comments. I fixed a bug while extending the test.

Regarding your comment about `SupportsProjectionPushDown` vs `SupportsMetadataReading`: We can open a follow-up issue for the pruning of unused metadata columns. This should not have big performance impart because usually metadata is only declared in tables when it is also needed.

wuchong commented:
I'm fine with a follow-up issue for pruning metadata columns.

wuchong reviewed (approved):
LGTM.

+1 to merge once build turns green.
